

strainNameForStrain = [""] * 37
genomeFileNameForStrain = [""] * 37
rnaSeqFileNameForStrain = [""] * 37
motifsForStrain = [[]] * 37
## Motifs are Based on 2018.11.29_SC FRosconi_Spneumoniae.xlsx. Only one methylation per motif


strainID = 1
strainNameForStrain[strainID] = "PG1"
genomeFileNameForStrain[strainID] = "CP035265.gbk"
rnaSeqFileNameForStrain[strainID] = "PG1_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GCATC", [2], 1), 
    ("GATGC", [1], 1), 
    ("TCTAGA", [5], 1), 
    ("CRAANNNNNNNNNTTC", [3], 1), 
    ("CRAANNNNNNNNCTK", [3], 1), 
    ("MAGNNNNNNNNTTYG", [1], 1),  
    ("GAAHNNNNNNNNTTYG", [2], 1)
    ]


strainID = 2
strainNameForStrain[strainID] = "PG2"
genomeFileNameForStrain[strainID] = "CP035264.gbk"
rnaSeqFileNameForStrain[strainID] = "PG2_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATGC", [1], 1), 
    ("GCATC", [2], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 3
strainNameForStrain[strainID] = "PG3"
genomeFileNameForStrain[strainID] = "CP035263.gbk"
rnaSeqFileNameForStrain[strainID] = "PG3_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GAYNNNNNCTRC", [1], 1),
    ("MAGNNNNNNNNTTYG", [1], 1),
    ("GYAGNNNNNRTC", [2], 1),
    ("TCTAGA", [5], 1),
    ("CRAANNNNNNNNCTK", [3], 1)
    ]


strainID = 4
strainNameForStrain[strainID] = "PG4"
genomeFileNameForStrain[strainID] = "CP035262.gbk"
rnaSeqFileNameForStrain[strainID] = "PG4_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1),
    ("GAYNNNNNNTATC", [1], 1),
    ("GATANNNNNNRTC", [3], 1),
    ("AAGNNNNNNNNTTYG", [1], 1),
    ("CRAANNNNNNNNCTT", [3], 1)
    ]


strainID = 5
strainNameForStrain[strainID] = "PG5"
genomeFileNameForStrain[strainID] = "CP035261.gbk"
rnaSeqFileNameForStrain[strainID] = "PG5_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATC", [1], 1),
    ("TCTAGA", [5], 1),
    ("CAGNNNNNNNGTG", [1], 1),
    ("CACNNNNNNNNTTC", [1], 1),
    ("GAANNNNNNNNGTG", [2], 1),
    ("CACNNNNNNNCTG", [1], 1)
    ]


strainID = 6
strainNameForStrain[strainID] = "PG6"
genomeFileNameForStrain[strainID] = "CP035260.gbk"
rnaSeqFileNameForStrain[strainID] = "PG6_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATC", [1], 1),
    ("TCTAGA", [5], 1),
    ("MAGNNNNNNNNTTYG", [1], 1),
    ("CRAANNNNNNNNCTK", [3], 1)
    ]


strainID = 7
strainNameForStrain[strainID] = "PG7"
genomeFileNameForStrain[strainID] = "CP035259.gbk"
rnaSeqFileNameForStrain[strainID] = "PG7_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATANNNNNNNTCA", [3], 1),
    ("TGANNNNNNNTATC", [2], 1)
    ]


strainID = 8
strainNameForStrain[strainID] = "PG8"
genomeFileNameForStrain[strainID] = "CP035258.gbk"
rnaSeqFileNameForStrain[strainID] = "PG8_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1),
    ("GTAYNNNNNNTCA", [2], 1),
    ("TGANNNNNNRTAC", [2], 1),
    ("GAANNNNNNNNNTTYG", [2], 1),
    ("CRAANNNNNNNNNTTC", [3], 1),
    ("CAGNNNNNNNNTTYG", [1], 1)
    ]


strainID = 9
strainNameForStrain[strainID] = "PG9"
genomeFileNameForStrain[strainID] = "CP035257.gbk"
rnaSeqFileNameForStrain[strainID] = "PG9_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1)
    ]


strainID = 10
strainNameForStrain[strainID] = "PG10"
genomeFileNameForStrain[strainID] = "CP038253.gbk"
rnaSeqFileNameForStrain[strainID] = "PG10_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1),
    ("GATANNNNNNRTC", [3], 1),
    ("GAYNNNNNNTATC", [1], 1)
    ]


strainID = 11
strainNameForStrain[strainID] = "PG11"
genomeFileNameForStrain[strainID] = "CP035256.gbk"
rnaSeqFileNameForStrain[strainID] = "PG11_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATC", [1], 1),
    ("TCTAGA", [5], 1),
    ("GTAYNNNNNNTCA", [2], 1),
    ("TGANNNNNNRTAC", [2], 1)
    ]


strainID = 12
strainNameForStrain[strainID] = "PG12"
genomeFileNameForStrain[strainID] = "CP035255.gbk"
rnaSeqFileNameForStrain[strainID] = "PG12_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATC", [1], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 13
strainNameForStrain[strainID] = "PG13"
genomeFileNameForStrain[strainID] = "CP035254.gbk"
rnaSeqFileNameForStrain[strainID] = "PG13_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("MAGNNNNNNNNTTYG", [1], 1),
    ("TCTAGA", [5], 1),
    ("CRAANNNNNNNNCTK", [3], 1)
    ]


strainID = 14
strainNameForStrain[strainID] = "PG14"
genomeFileNameForStrain[strainID] = "CP035253.gbk"
rnaSeqFileNameForStrain[strainID] = "PG14_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1)
    ]


strainID = 15
strainNameForStrain[strainID] = "PG15"
genomeFileNameForStrain[strainID] = "CP035252.gbk"
rnaSeqFileNameForStrain[strainID] = "PG15_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GAYNNNNNNTATC", [1], 1),
    ("GATANNNNNNRTC", [3], 1),
    ("TCTAGA", [5], 1),
    ("CAGNNNNNNNGTG", [1], 1),
    ("CACNNNNNNNCTG", [1], 1)
    ]


strainID = 16
strainNameForStrain[strainID] = "PG16"
genomeFileNameForStrain[strainID] = "CP035251.gbk"
rnaSeqFileNameForStrain[strainID] = "PG16_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCGAG", [3], 1),
    ("GAYNNNNNNTATC", [1], 1),
    ("GATANNNNNNRTC", [3], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 17
strainNameForStrain[strainID] = "PG17"
genomeFileNameForStrain[strainID] = "CP038252.gbk"
rnaSeqFileNameForStrain[strainID] = "PG17_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCGAG", [3], 1),
    ("GAYNNNNNCTRC", [1], 1),
    ("GYAGNNNNNRTC", [2], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 18
strainNameForStrain[strainID] = "PG18"
genomeFileNameForStrain[strainID] = "CP035250.gbk"
rnaSeqFileNameForStrain[strainID] = "PG18_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCGAG", [3], 1),
    ("TCTAGA", [5], 1),
    ("GATANNNNNNRTC", [3], 1),
    ("GAYNNNNNNTATC", [1], 1)
    ]


strainID = 19
strainNameForStrain[strainID] = "PG19"
genomeFileNameForStrain[strainID] = "CP035249.gbk"
rnaSeqFileNameForStrain[strainID] = "PG19_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATC", [1], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 20
strainNameForStrain[strainID] = "PG20"
genomeFileNameForStrain[strainID] = "CP035248.gbk"
rnaSeqFileNameForStrain[strainID] = "PG20_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATC", [1], 1),
    ("GAYNNNNNNTCC", [1], 1),
    ("GGANNNNNNRTC", [2], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 21
strainNameForStrain[strainID] = "PG21"
genomeFileNameForStrain[strainID] = "CP035247.gbk"
rnaSeqFileNameForStrain[strainID] = "PG21_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GAYNNNNNNTATC", [1], 1),
    ("GATANNNNNNRTC", [3], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 22
strainNameForStrain[strainID] = "PG22"
genomeFileNameForStrain[strainID] = "CP035246.gbk"
rnaSeqFileNameForStrain[strainID] = "PG22_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GAYNNNNNNTATC", [1], 1),
    ("GATANNNNNNRTC", [3], 1),
    ("TCTAGA", [5], 1),
    ("GAANNNNNNNNNTTYG", [2], 1),
    ("CRAANNNNNNNNNTTC", [3], 1)
    ]


strainID = 23
strainNameForStrain[strainID] = "PG23"
genomeFileNameForStrain[strainID] = "SCPS00000000.gbk"                                      
rnaSeqFileNameForStrain[strainID] = "PG23_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1)
    ]


strainID = 24
strainNameForStrain[strainID] = "PG24"
genomeFileNameForStrain[strainID] = "CP035245.gbk"
rnaSeqFileNameForStrain[strainID] = "PG24_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1)
    ]


strainID = 25
strainNameForStrain[strainID] = "PG25"
genomeFileNameForStrain[strainID] = "CP038251.gbk"
rnaSeqFileNameForStrain[strainID] = "PG25_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("MAGNNNNNNNNTTYG", [1], 1)
    ]


strainID = 26
strainNameForStrain[strainID] = "PG26"
genomeFileNameForStrain[strainID] = "CP035244.gbk"
rnaSeqFileNameForStrain[strainID] = "PG26_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1)
    ]


strainID = 27
strainNameForStrain[strainID] = "PG27"
genomeFileNameForStrain[strainID] = "CP035243.gbk"
rnaSeqFileNameForStrain[strainID] = "PG27_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ]


strainID = 28
strainNameForStrain[strainID] = "PG28"
genomeFileNameForStrain[strainID] = "CP035242.gbk"
rnaSeqFileNameForStrain[strainID] = "PG28_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GAYNNNNNNTCC", [1], 1),
    ("TCGAG", [3], 1),
    ("GGANNNNNNRTC", [2], 1),
    ("TCTAGA", [5], 1),
    ("CAGNNNNNNNGTG", [1], 1),
    ("CACNNNNNNNNTTC", [1], 1),
    ("GAANNNNNNNNGTG", [2], 1),
    ("CACNNNNNNNCTG", [1], 1)
    ]


strainID = 29
strainNameForStrain[strainID] = "PG29"
genomeFileNameForStrain[strainID] = "CP035241.gbk"
rnaSeqFileNameForStrain[strainID] = "PG29_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("GATC", [1], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 30
strainNameForStrain[strainID] = "PG30"
genomeFileNameForStrain[strainID] = "CP035240.gbk"                                              #############################################################################
rnaSeqFileNameForStrain[strainID] = "PG30_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCGAG", [3], 1),
    ("AAGNNNNNNNNTTYG", [1], 1),
    ("TCTAGA", [5], 1),
    ("CRAANNNNNNNNCTT", [3], 1)
    ]


strainID = 31
strainNameForStrain[strainID] = "CT-22F"
genomeFileNameForStrain[strainID] = "CP035235.gbk"
rnaSeqFileNameForStrain[strainID] = "CT22F_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1),
    ("CAGNNNNNNNNTTYG", [1], 1),
    ("CRAANNNNNNNNCTG", [3], 1)
    ]


strainID = 32
strainNameForStrain[strainID] = "GA-22F"
genomeFileNameForStrain[strainID] = "CP035234.gbk"
rnaSeqFileNameForStrain[strainID] = "GA22F_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCTAGA", [5], 1),
    ("GAANNNNNNNNNTTYG", [2], 1),
    ("CRAANNNNNNNNNTTC", [3], 1)
    ]


strainID = 33
strainNameForStrain[strainID] = "D39"
genomeFileNameForStrain[strainID] = "CP035238.gbk"
rnaSeqFileNameForStrain[strainID] = "D39_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCGAG", [3], 1),
    ("TCTAGA", [5], 1),
    ("GAANNNNNNNNNTTYG", [2], 1),
    ("CRAANNNNNNNNNTTC", [3], 1)
    ]


strainID = 34
strainNameForStrain[strainID] = "TIGR4"
genomeFileNameForStrain[strainID] = "CP035239.gbk"
rnaSeqFileNameForStrain[strainID] = "TIGR4_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCGAG", [3], 1),
    ("GAYNNNNNNTATC", [1], 1),
    ("GATANNNNNNRTC", [3], 1),
    ("TCTAGA", [5], 1),
    ("MAGNNNNNNNNTTYG", [1], 1),
    ("CRAANNNNNNNNCTT", [3], 1)
    ]


strainID = 35
strainNameForStrain[strainID] = "Taiwan19F"
genomeFileNameForStrain[strainID] = "CP035237.gbk"
rnaSeqFileNameForStrain[strainID] = "Taiwan19F_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ("TCANNNNNNNTCC", [2], 1),
    ("GGANNNNNNNTGA", [2], 1),
    ("TCTAGA", [5], 1)
    ]


strainID = 36
strainNameForStrain[strainID] = "BHN97"
genomeFileNameForStrain[strainID] = "CP035236.gbk"
rnaSeqFileNameForStrain[strainID] = "BHN97_RNAseq.xlsx"
motifsForStrain[strainID] = [
    ]